### PR TITLE
Add Projects API client (preview)

### LINKED ISSUE #1472: Support Projects preview API
As per the [recent anouncement](https://developer.github.com/changes/2016-09-14-projects-api/) there is a new API for the Projects functionality

[API docs](https://developer.github.com/v3/repos/projects/)

There are 17 functions to be implemented around Projects, Columns and Cards.  It's also likely that there are `repositoryId` based versions of these functions as well

### PR DESCRIPTION
**Pull request by maddin2016**
fix #1472

Add support for the new Projects API

### CODE REVIEW

**Comment by maddin2016** [comment]
Hey @ryangribble. I try to `FixProjects` and get this
```
build.fsx(120,14): error FS0039: The value, constructor, namespace or type 'DotNetCli' is not defined

CACHING WARNING
this might happen after Updates...
please open a issue on FAKE and /cc @matthid ONLY IF this happens reproducibly)

Error: System.Exception: We could not find a type similar to '<StartupCode$FSI_0001>.$FSI_0001_Build$fsx' containing a 'main@' method in the cached assembly (./.fake/build.fsx_857B7771.dll)!
   bei Fake.FSIHelper.runScriptCached@328-4.Invoke(String message) in C:\code\fake\src\app\FakeLib\FSIHelper.fs:Zeile 328.
   bei Yaaf.FSharp.Scripting.Helper.consoleCapture[a](TextWriter out, TextWriter err, FSharpFunc`2 f) in C:\code\fake\paket-files\matthid\Yaaf.FSharp.Scripting\src\source\Yaaf.FSharp.Scripting\YaafFSharpScripting.fs:Zeile 1018.
   bei Fake.FSIHelper.runScriptCached(Boolean printDetails, CacheInfo cacheInfo, TextWriter out, TextWriter err) in C:\code\fake\src\app\FakeLib\FSIHelper.fs:Zeile 313.
   bei Fake.FSIHelper.runFAKEScriptWithFsiArgsAndRedirectMessages(Boolean printDetails, FsiArgs _arg1, IEnumerable`1 env, FSharpFunc`2 onErrMsg, FSharpFunc`2 onOutMsg, Boolean useCache) in C:\code\fake\src\app\FakeLib\FSIHelper.fs:Zeile 512.
Unable to access C:\Users\scholzmartin\Documents\GitHub\octokit.net\.fake\build.fsx_857B7771.dll
FsiEvaluationException:

Error:

        build.fsx(120,14): error FS0039: The value, constructor, namespace or type 'DotNetCli' is not defined


Output: [Loading C:\Users\scholzmartin\Documents\GitHub\octokit.net\tools\SourceLink.Fake\tools\SourceLink.fsx
         Loading C:\Users\scholzmartin\Documents\GitHub\octokit.net\build.fsx]
```

**Comment by ryangribble** [comment]
How are you executing FixProjects?  Are you up to date with latest master?  

[This change](https://github.com/octokit/octokit.net/commit/091f9bdb814e633909e279f01aa6b7cc2f4332be) to `build.cmd` bootstrap script updated the nuget restore command to grab a later version of FAKE that brought dotnet core tasks.  It was 14 days ago so I assume you already must have that commit, but from the error message you're getting it looks like FAKE doesn't recognize `DotNetCli` tasks which would indicate you arent on that later version that the bootstrap script should be updating to?

**Comment by maddin2016** [comment]
Mmhh, but i'm currently up to date with master
image

**Comment by maddin2016** [comment]
i have looked in tools folder an it seems that i still have fake version 4.28 installed. How can get update fake. Normaly `build.cmd` should do this, right?

**Comment by maddin2016** [comment]
Ok, i've got it. I have renamed the `FAKE.Core` folder in tools and then run the build script again. Now it works. It seems that `nuget.exe` doesn't override the existing version.

**Comment by maddin2016** [comment]
mmh, but now there are untracked files for the `FAKE.Core` Folder. How can i exclude this folder?

**Comment by maddin2016** [comment]
I don't have delete the renamed folder 😸

**Comment by maddin2016** [comment]
Hi @ryangribble, how would we named [this](https://developer.github.com/v3/repos/projects/#move-a-column). 
`RepositoryProjectColumnMove`??

**Comment by maddin2016** [comment]
How do we construct classes where we have two parameters like [here](https://developer.github.com/v3/repos/projects/#create-a-project-card) and we have to provide one of them? Set the first one as required?

**Comment by maddin2016** [comment]
Should i create multiple constructors
```
public NewRepositoryProjectCard(string note)
        {
            Note = note;
        }

public NewRepositoryProjectCard(int contentId)
        {
            ContentId = contentId;
        }

public string Note { get; protected set; }

public int ContentId { get; protected set; }

```

**Comment by ryangribble** [comment]
Naming things is always hard 😀 what you've suggested sounds fine 

Although you probably know me well enough to know that I always have an opinion :grinning: so if discussion is wanted over naming of the request/response model classes, my 2c is that "Project"  is already a sufficient enough entity to name everything with so I probably would use "Project" "ProjectCard" "ProjectColumn" rather than "RepositoryProjectXxx".  For example a pull request, issue etc all belong to a repository but we don't call them "RepositoryPullRequest" "RepositoryIssue" 

That's just my thoughts, not saying you have to rename everything (unless you agree with my assertions)

File: Octokit/Helpers/AcceptHeaders.cs [hunk 1]
```diff
@@ -42,5 +42,7 @@ public static class AcceptHeaders
         public const string IssueTimelineApiPreview = "application/vnd.github.mockingbird-preview";
 
         public const string RepositoryTrafficApiPreview = "application/vnd.github.spiderman-preview";
+
+        public const string ProjectsApiPreview = "  application/vnd.github.inertia-preview+json";
```
Comment (ryangribble) [inline Octokit/Helpers/AcceptHeaders.cs]: Looks like leading whitespace in the string?

**Comment by maddin2016** [comment]
>Although you probably know me well enough to know that I always have an opinion 😀

I know 😄 😸 . But i wasn't sure because there are sometimes classes which be called different. For eg.
`PullRequestRequest.cs` and on the other side there is `RepositoryIssueRequest.cs`. So i agree with you that `Project`sounds better then `RepositoryProject`. So i think will rename them.

**Comment by maddin2016** [comment]
Or another examples are `RepositoryInvitation.cs`, `RepositoryTag.cs`, `RepositoryHook.cs`, and there are lot more, mmhhh 😕

**Comment by ryangribble** [comment]
Yeah I guess I can only vouch for things since I've been contributing so for anything that pre-dates that I reserve the right to say I wouldn't have named it that, haha 😂 

A note on hooks though - you can have repo or organisation level hooks so in the case at least, it's necessary to denote the type/level of hook. Infact invitations also might exist at both repo an org levels?

**Comment by maddin2016** [comment]
Hopefully no one reads this 😜 
Ok, that make sense. So i agree as long as there are only projects for n<=1 types we call them Project.

**Comment by maddin2016** [comment]
keyboard
Phew ... tests are finished. So i think it's time to review.

**Comment by ryangribble** [comment]
Hey @maddin2016 sorry for not getting stuck into this yet!

Not sure if you saw but some changes were recently announced regarding this preview API.  Do you think it just impacts your endpoints and function parameters or does it also mean the actual calls need to live in a different place?

[GitHub Developer Blog - Changes to Projects API](https://developer.github.com/changes/2016-10-27-changes-to-projects-api/)

**Comment by maddin2016** [comment]
Hi @ryangribble, wow, that's a lot of changes. So i have to step through. Please correct me if i'm wrong

1. There are 2 new endpoints `List organization projects` and `Create organization projects`

1. The responses of this two are same like for public projects. So i think they stay under the same client `RepositoryProjectsClient` and we name them `GetAllForOrganization` and `CreateForOrganization`

1. `/repos/:owner/:repo/projects/:number` endpoints are deprecated and will be removed on 10. November. So i think we should wait for 10. November before we add this feature into client. So we can  already remove these endpoints and only use the new  id endpoints `/projects/:id`

Or do you think we should add these changes now and remove them after 10. November? Do we have some work with scopes on client side?

**Comment by maddin2016** [comment]
Hey @ryangribble, it was a little late yesterday. I just now realize that they completly reorganized the projects api. So i think like you say they need to live in a different place.

1. I would create a client `ProjectsClient` with the api calls only for projects.

1. Then i would create two new seperated clients for project cards and project columns `ProjectCardsClient` and `ProjectColumnsClient`

What do you say?

**Comment by ryangribble** [comment]
Yep this aligns with th API docs structure so I think it's the way to go. I guess this (significant changes) is the  risk we run when implementing these preview APIs 😂

**Comment by maddin2016** [comment]
Yeap, but i think it's the right way to add these preview features instead of waiting until they are finished!! Ok then i will rewrite clients and hope that they don't switch back 😸 But there is still the question if we only implement the new endpoints `projects/.../:id`?

**Comment by ryangribble** [comment]
> But there is still the question if we only implement the new endpoints projects/.../:id?

Yeah definitely only do the new ones, since the old ones will be gone in the next couple of weeks anyway

**Comment by maddin2016** [comment]
image


happygif

🎉 🎉

**Comment by maddin2016** [comment]
So ... i have rewrite clients. Things to do is change the tests.

**Comment by maddin2016** [comment]
Soo, all is rewritten. 
Hopefully they don't change api again

fist
😸

**Comment by maddin2016** [comment]
Integration test's all have passed except the one for Organization. @ryangribble can you please test this test.

**Comment by ryangribble** [comment]
Hey @maddin2016 

Ive finally got around to pulling this down.  I had to make some changes to integration tests but other than that everything is looking :gem:

Any chance you can give me permission to push my change to your branch by ticking the box on the right? ----->

image

File: Octokit/Clients/IProjectsClient.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        /// <param name="id">The Id of the project</param>
+        Task<bool> Delete(int id);
+
+        /// <summary>
+        /// A client for GitHub's Project Cards API.
+        /// </summary>
+        /// <remarks>
+        /// See the <a href="https://developer.github.com/v3/repos/projects/cards/">Repository Projects API documentation</a> for more information.
+        /// </remarks>
+        IProjectCardsClient Cards { get; }
```
Comment (ryangribble) [inline Octokit/Clients/IProjectsClient.cs]: Cards => Card

File: Octokit/Clients/IProjectsClient.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+        /// <summary>
+        /// A client for GitHub's Project Columns API.
+        /// </summary>
+        /// <remarks>
+        /// See the <a href="https://developer.github.com/v3/repos/projects/columns/">Repository Projects API documentation</a> for more information.
+        /// </remarks>
+        IProjectColumnsClient Columns { get; }
```
Comment (ryangribble) [inline Octokit/Clients/IProjectsClient.cs]: Columns => Column

File: Octokit/Clients/IRepositoriesClient.cs [hunk 4]
```diff
@@ -663,5 +663,13 @@ public interface IRepositoriesClient
...
         /// </remarks>
         IRepositoryTrafficClient Traffic { get; }
+
+        /// <summary>
+        /// Access GitHub's Repository Projects API
+        /// </summary>
+        /// <remarks>
+        /// Refer to the API documentation for more information: https://developer.github.com/v3/repos/projects/
+        /// </remarks>
+        IProjectsClient Projects { get; }
```
Comment (ryangribble) [inline Octokit/Clients/IRepositoriesClient.cs]: Projects => Project

File: Octokit/Clients/RepositoriesClient.cs [hunk 5]
```diff
@@ -966,5 +967,13 @@ public Task<Branch> GetBranch(string owner, string name, string branchName)
...
         /// </remarks>
         public IRepositoryTrafficClient Traffic { get; private set; }
+
+        /// <summary>
+        /// Access GitHub's Repository Projects API
+        /// </summary>
+        /// <remarks>
+        /// Refer to the API documentation for more information: https://developer.github.com/v3/repos/projects/
+        /// </remarks>
+        public IProjectsClient Projects { get; private set; }
```
Comment (ryangribble) [inline Octokit/Clients/RepositoriesClient.cs]: Projects => Project

File: Octokit/Clients/ProjectsClient.cs [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        /// </remarks>
+        public IProjectCardsClient Cards { get; private set; }
+
+        /// <summary>
+        /// A client for GitHub's Project Columns API.
+        /// </summary>
+        /// <remarks>
+        /// See the <a href="https://developer.github.com/v3/repos/projects/columns/">Repository Projects API documentation</a> for more information.
+        /// </remarks>
+        public IProjectColumnsClient Columns { get; private set; }
```
Comment (ryangribble) [inline Octokit/Clients/ProjectsClient.cs]: Columns => Column

File: Octokit/Clients/ProjectsClient.cs [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+            }
+        }
+
+        /// <summary>
+        /// A client for GitHub's Project Cards API.
+        /// </summary>
+        /// <remarks>
+        /// See the <a href="https://developer.github.com/v3/repos/projects/cards/">Repository Projects API documentation</a> for more information.
+        /// </remarks>
+        public IProjectCardsClient Cards { get; private set; }
```
Comment (ryangribble) [inline Octokit/Clients/ProjectsClient.cs]: Cards => Card

**Comment by maddin2016** [comment]
@ryangribble, `Edit from maintainers` is enabled

**Comment by markwilkie** [comment]
I found an additional problem when adding an issue card to a project column using projectCardsClient.Create.  GitHub wants the word 'Issue', not 'issue' for context-type.  (https://developer.github.com/v3/projects/cards/#create-a-project-card)  However, in EnumExtensions, it ToLowers the property.  I'm not 100% sure the ramifications, so I'm not submitting a fix.  However, I can confirm that when I remove the ToLower, I can add a card.

**Comment by maddin2016** [comment]
@markwilkie, thanks for that tip. I can confirm that is if 'issue' is lowercase it returns an error. @ryangribble, @shiftkey do you think thats an error by the api? Because in all other api such parameters are lowcase.

**Comment by maddin2016** [comment]
I have set `Parameter(Value="Issue")`. Now client works.

**Comment by maddin2016** [comment]
@ryangribble, do you have the chance to push your changes for integration tests. Because i forgot to test to create a card from an issue.

**Comment by ryangribble** [comment]
OK ive pushed up the test changes and also updated to latest master

To explain what i did with the tests:  There is no need to use GHE stuff for the tests as we dont need to create an organization, we can just create projects in the test user's existing organization.  Also had to tweak some of the asserts to only ensure the created projects existed in the returned values rather than the count being 2 (since existing projects may cause there to be more than 2).  I also simplified the helper functions to not pass around an entire repository context object but just the repository Id, and finally fixed a couple of bugs where repository name was being used instead of organization name :+1:

All integration tests that are there are passing for me (and you should be able to run them all too) but I guess as you've pointed out, there still appear to be some methods that dont have tests

**Comment by maddin2016** [comment]
Hey @ryangribble, happy new year 🎉 Seems this is ready to 🚢 ?

**Comment by ryangribble** [comment]
Yes, I merged in latest `master` and will just do a final review and integration test run tonight, then it can be merged :tada:

**Comment by ryangribble** [comment]
Hey @maddin2016 did you check if the Projects API supports pagination?  Currently there are no `GetAll` methods implemented with `ApiOptions` overloads

**Comment by shiftkey** [comment]
@ryangribble @maddin2016 the documentation seems to suggest this is supported :wink:

**Comment by ryangribble** [comment]
Soooo close!!!!



If you can add the `ApiOptions` overloads please @maddin2016 then we can merge it in 😀

**Comment by maddin2016** [comment]
I also added `ApiOptions` for `ProjectCards` and `ProjectColumns`.

**Comment by ryangribble** [comment]
Changes look good.  We also normally have integration tests that prove the pagination is working - such as [these ones](https://github.com/octokit/octokit.net/blob/master/Octokit.Tests.Integration/Clients/PullRequestsClientTests.cs#L78-L222) for the PullRequest client

**Comment by ryangribble** [comment]
@maddin2016 Im putting together a new release shortly, just wondering if you want some help with the integration tests for the pagination?

**Comment by maddin2016** [comment]
Hey @ryangribble, I'm currently on the road. Tomorrow I add the missing test's so that we quickly can close this OR

**Comment by shaggygi** [comment]
@maddin2016 @ryangribble any updates when this will be added? Just curious.  Thx

**Comment by ryangribble** [comment]
I thought we were still waiting for more integration tests to be added... I'll need to have a look

**Comment by ryangribble** [comment]
Hey so I think ive tidied this one up.  While I was adding the missing pagination tests, I actually discovered there were some new methods and fields that hadnt been implemented back when this was first started

All the tests are passing, including integration tests but it'd be great if someone can take a sanity check 👀 before I hit the button?  @maddin2016 @mderriey @M-Zuber

**Comment by M-Zuber** [comment]
Looks good to me from what I can tell

**Comment by mderriey** [comment]
Ditto. Great work getting this back on track :+1:

**Comment by ryangribble** [comment]
Thanks @maddin2016 for the ground work on this one!

LGTM

**Comment by ryangribble** [comment]
release_notes:  Implement support for [Projects API (Preview)](https://developer.github.com/v3/projects/)

**Pull request merged**